src/main/groovy/org/assertj/generator/gradle/tasks/config/Templates.groovy

```diff
@@ -33,248 +37,261 @@ import static org.assertj.assertions.generator.Template.Type.*
+            }
+            templateFiles = objects.fileCollection().tap {
+                from(templateInfos.map { infos -> infos.collect { it.fileOrTemplate.file.getOrNull() } })
+            }
         }
 
-        void writeOutputStream(ObjectOutputStream s) throws IOException {
-            this.metaClass.properties.findAll { prop ->
-                Either.class.isAssignableFrom(prop.type)
-            }.each {
-                s.writeObject(it.getProperty(this))
+        @EqualsAndHashCode
+        @ToString
+        protected static class Info {
+            final FileOrTemplate fileOrTemplate
+            final Template.Type type
+            final String description
+
+            SerializedTemplate toTemplate() {
+                if (fileOrTemplate.file.isPresent()) {
+                    def file = fileOrTemplate.file.asFile.get()
+                    new SerializedTemplate(type, null, file)
+                } else if (fileOrTemplate.template.isPresent()) {
+                    new SerializedTemplate(type, fileOrTemplate.template.get(), null)
+                } else {
+                    null
+                }
             }
-        }
 
-        void fromInputStream(ObjectInputStream s) throws IOException, ClassNotFoundException {
-            this.metaClass.properties.findAll { prop ->
-                Either.class.isAssignableFrom(prop.type)
-            }.each { prop ->
-                prop.setProperty(this, s.readObject())
+            private Info(
+                    FileOrTemplate fileOrTemplate,
+                    Template.Type type,
+                    String description
+            ) {
+                this.fileOrTemplate = fileOrTemplate
+                this.type = type
+                this.description = description
             }
         }
 
+        static protected Provider<Info> info(final Property<FileOrTemplate> self, Template.Type type, String description) {
+            self.map { new Info(it, type, description) }
+        }
     }
 
     /**
      * Class-level templates
      */
     @EqualsAndHashCode
-    class ClassTemplates extends TemplateHandler implements Serializable {
-
-        Either<File, String> assertionClass
-        Either<File, String> hierarchicalConcrete
-        Either<File, String> hierarchicalAbstract
-
-        @Override
-        def getTemplates(List<Template> userTemplates, AssertionsGeneratorReport report) {
-            // @format:off
-            loadUserTemplate(assertionClass, ASSERT_CLASS, "'class assertions'", userTemplates, report)
-            loadUserTemplate(hierarchicalConcrete, HIERARCHICAL_ASSERT_CLASS, "'hierarchical concrete class assertions'", userTemplates, report)
-            loadUserTemplate(hierarchicalAbstract, ABSTRACT_ASSERT_CLASS, "'hierarchical abstract class assertions'", userTemplates, report)
-            // @format:on
+    static class ClassTemplates extends TemplateHandler {
```

Comment:
None of the content of these classes changed, only the registration mechanism was simplified